### PR TITLE
Add a --lint flag to the promtool check rules and check config commands

### LINKED ISSUE #8613: promtool: Support some kind of “lint” for rules
`promtool check rules` currently is [documented](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#syntax-checking-rules) as a way to syntax check rules. However it also checks for duplicate rules, which is closer to a lint check.

There are some features in PromQL that are potentially problematic to use in recording rules now:
- Negative offset is likely incorrect in recording rules (I mentioned some kind of lint [then](https://github.com/prometheus/prometheus/issues/6282#issuecomment-550308545))
- Subqueries are often not a good idea (noted on the [blog post](https://prometheus.io/blog/2019/01/28/subquery-support/))

I propose something like a `--lint` flag where the default is “duplicate-rules” to match current behaviour, but could also accept `--lint=all`, etc. 

(Some overlap with #8607)

### PR DESCRIPTION
**Pull request by fpetkovski**
Checking rules with promtool emits warnings in the case of duplicate rules.
These warnings do not result in a non-zero exit code and are difficult to
spot in CI environments. Additionally, checking for duplicates is closer
to a lint check rather than a syntax check.

This commit adds a --lint flag to commands which include checking rules.
The flag can be used to enable or disable certain linting options
and cause the execution to return a non-zero exit code in case
those options are not met.

Fixes #8613

Signed-off-by: fpetkovski <filip.petkovsky@gmail.com>

### CODE REVIEW

File: cmd/promtool/main.go
Comment (LeviHarrison) [inline cmd/promtool/main.go:296]: Instead of storing these at strings, I think it might be simpler to handle them like we do feature flags, which is a struct containing booleans for all the options.

File: cmd/promtool/main.go
Comment (LeviHarrison) [inline cmd/promtool/main.go:604]: > cause the execution to return a non-zero exit code in case
those options are not met.

So should this be a non-zero exit code, or is this a lint check that still returns a success?

**Review by LeviHarrison — COMMENTED** [review]
Thanks!

File: cmd/promtool/main.go
Comment (fpetkovski) [reply cmd/promtool/main.go:604]: I used the same pattern as in the error condition above, but there is also a `lintErrExitCode` that we can return.
https://github.com/prometheus/prometheus/blob/53ac9d6d666acaebbdb8a004b51958f3147b1bd5/cmd/promtool/main.go#L548-L551

However, it seems that this value is ignored when errors are returned from `checkRules` because it indicates how many rules are discovered, it's not the actual exit code:
https://github.com/prometheus/prometheus/blob/53ac9d6d666acaebbdb8a004b51958f3147b1bd5/cmd/promtool/main.go#L303-L319

So maybe we should just use `0` in the case the linter fails?

Comment (fpetkovski) [reply cmd/promtool/main.go]: Looking further, `lintErrExitCode` is only used in `CheckMetrics`. Do we want to use it in the rules linter as part of this PR?

Comment (LeviHarrison) [reply cmd/promtool/main.go]: Thanks! I like the approach.

File: cmd/promtool/main.go
Comment (LeviHarrison) [inline cmd/promtool/main.go:300]: nit: because this is printed in the output at the start of the linting errors, it should be capitalized

```suggestion
var lintError = fmt.Errorf("Lint error")
```

File: cmd/promtool/main.go
Comment (LeviHarrison) [inline cmd/promtool/main.go:625]: nit: same thing here, just to improve presentation

```suggestion
			return 0, []error{fmt.Errorf("%w: %s", lintError, errMessage)}
```

File: cmd/promtool/main.go
Comment (LeviHarrison) [inline cmd/promtool/main.go:300]: IMO this naming makes more sense than what revive wants, so we can ignore it.

```suggestion
// nolint:revive
var lintError = fmt.Errorf("lint error")
```

**Review by LeviHarrison — APPROVED** [review]
Looks great, thanks! A rebase should fix the `test_windows` failure.

I'll leave this around for a few days before merging in case any of the Promtool maintainers have additional feedback.

**Comment by fpetkovski** [comment]
Thanks for the review. Would you like me to do the rebase on this branch, or would that happen automatically when merging to main?

**Comment by Nexucis** [comment]
@fpetkovski yeah can you rebase your PR on this branch please ?

It's just to get back the changes from the main in your branch and see if the CI job `test_windows` can be green.

By the way, in two days I will cut a new release (v2.35). So you could either try to expedite this enough to go into the next release. Or, on the other hand, if you feel this is too invasive to go into a release without sitting a while in main, you might want to delay the merge.

**Comment by mogoman** [comment]
Hi. Just a quick question here; running v2.35 of promtool now exits for us with code 3 as we have duplicate rule names defined - which makes sense from the changes here. However Prometheus v2.35 still starts with duplicate rules - is this correct, or should Prometheus also reject duplicate rules?

**Comment by LeviHarrison** [comment]
Yes, I believe that is expected behavior. Duplicate rules are completely fine, they're just something that's rarely actually intended, hence it being a lint check.

**Comment by Sebi94nbg** [comment]
Hello!
Regarding @mogoman question: We have the same problem. We had to disable the linting option "duplicate-rules" as it fails since this lint option is included in the promtool.

@LeviHarrison you already wrote, that it's completely fine as it's intended like this to have "duplicate" rules. But will this be deprecated (and unsupported) in the future, so that we have to fix this issue?

Is there anywhere a documentation with best practices how alerting rules should be defined incl. such information how you can avoid "duplicate rules"? We for example have multiple alerts with the same alert rule name, but with different thresholds, severities and sometimes even `for`. The reason for it is, that we wanted to overwrite active alerts with a low severity with a higher severity.

Thank you in advance!

**Comment by dgl** [comment]
Hi @Sebi94nbg -- for the current problem please upgrade promtool to at least 2.36.1, we've fixed the error status in that release.

I'm afraid there aren't very good docs on this -- we have #7301 open to track some improvements. With alert rules usually this is dealt with by having different labels identifying the alerts; something like:

```yaml
groups:
  - name: example
    rules:
      - alert: JobDown
        expr: my_warning_query
        labels:
          severity: warning
      - alert: JobDown
        expr: my_critical_query
        labels:
          severity: critical
```

The issue being alerts are identified by the complete set of labels, so if two alerts have exactly the same labels it is possible for strange things to happen (you mention `for` and that is one potential place where duplicated rules can cause alerts to resolve when one rule is still firing, etc). However in many cases this isn't a problem as labels flow through from the metrics (which the lint currently doesn't take into account).

**Comment by roidelapluie** [comment]
Duplicate rules are a fine use case. Not sure how to deal with them yet.

**Pull request merged**
